src/bindings/swift/mod.rs

```diff
@@ -0,0 +1,179 @@
+
+    let mut l =
+        File::create(&source_file).map_err(|e| anyhow!("Failed to create .swift file: {:?}", e))?;
+    write!(l, "{}", library).map_err(|e| anyhow!("Failed to write Swift library: {:?}", e))?;
+
+    Ok(())
+}
+
+/// ...
+pub fn compile_swift_module(ci: &ComponentInterface, out_dir: &OsStr) -> Result<()> {
+    let out_path = PathBuf::from(out_dir);
+
+    let mut module_map_file = out_path.clone();
+    module_map_file.push(format!("uniffi_{}.modulemap", ci.namespace()));
+
+    let mut module_map_file_option = OsString::from("-fmodule-map-file=");
+    module_map_file_option.push(module_map_file.as_os_str());
+
+    let mut source_file = out_path.clone();
+    source_file.push(format!("{}.swift", ci.namespace()));
+
+    let mut dylib_file = out_path.clone();
+    dylib_file.push(format!("lib{}.dylib", ci.namespace()));
+
+    // `-emit-library -o <path>` generates a `.dylib`, so that we can use the
+    // Swift module from the REPL. Otherwise, we'll get "Couldn't lookup
+    // symbols" when we try to import the module.
+    // See https://bugs.swift.org/browse/SR-1191.
+
+    let status = std::process::Command::new("swiftc")
+        .arg("-module-name")
+        .arg(ci.namespace())
+        .arg("-emit-library")
+        .arg("-o")
+        .arg(&dylib_file)
+        .arg("-emit-module")
+        .arg("-emit-module-path")
+        .arg(&out_path)
+        .arg("-parse-as-library")
+        .arg("-L")
+        .arg(&out_path)
+        .arg(format!("-luniffi_{}", ci.namespace()))
+        .arg("-Xcc")
+        .arg(module_map_file_option)
+        .arg(source_file)
+        .spawn()
+        .map_err(|_| anyhow::anyhow!("failed to spawn `swiftc`"))?
+        .wait()
+        .map_err(|_| anyhow::anyhow!("failed to wait for `swiftc` subprocess"))?;
+    if !status.success() {
+        bail!("running `swiftc` failed")
+    }
+    Ok(())
+}
+
+pub fn run_swift_script(out_dir: Option<&OsStr>, script_file: Option<&str>) -> Result<()> {
+    // TODO: Don't hard-code library names for the REPL.
+    const LIBS: &[&str] = &["-larithmetic", "-lgeometry", "-lsprites"];
+    const MODULE_MAPS: &[&str] = &[
+        "uniffi_arithmetic.modulemap",
+        "uniffi_geometry.modulemap",
+        "uniffi_sprites.modulemap",
+    ];
+
+    let mut cmd = std::process::Command::new("swift");
+    cmd.args(LIBS);
+
+    if let Some(out_dir) = out_dir {
+        let out_path = PathBuf::from(out_dir);
+        let cc_options: Vec<OsString> = MODULE_MAPS
+            .iter()
+            .flat_map(|module_map| {
+                let mut module_map_file = out_path.clone();
+                module_map_file.push(module_map);
+                if !module_map_file.exists() {
+                    // Missing module maps (for example, `uniffi_arithmetic` when we're
+                    // running the `geometry` example) will fail, so ignore maps that
+                    // don't exist. Gross!
```

Comment:
Hrm, this does seem fiddly. I wonder if you could do the reverse, and iterate over all files in `out_path` looking for any with the right extension, then deciding what to use based on that.

Reply:
That’s a great idea—I could have it search `out_path` for `.dylib`s and Swift module files (or, better yet, output the generated stuff into `swift`, `kotlin`, `python`, etc. subdirectories in `out_path`—so we’d avoid, say, linking the libs for `example`—or just add a filter for those). A special target subdirectory per language would also give you a good idea of what to include.

TBH, the only reason for this jiggery-pokery is so that you can run the Swift examples from the command line, or in a REPL. If you’re building with Xcode or SPM, you only need the bridging header and generated `.swift` file (SPM will make a `modulemap` for you, and also automatically make a dylib for the REPL—looking at that was how I figured out the right set of flags to pass to `swift` and `swiftc`). We only need the `swiftmodule`, `modulemap`, and `dylib` for `cargo run — exec`.

Reply:
Something else that might be useful here is, when each example component runs the command-line helper, it passes in the name of the component [here](https://github.com/rfk/uniffi-rs/blob/f28adde220adfcb01c2732b23d7c857010a0d268/src/lib.rs#L185). You might be able to thread that name through down to the call to `run_swift_script` and use it to help select which files to include here.